pandas/types/cast.py

```diff
@@ -322,8 +325,52 @@ def infer_dtype_from_scalar(val, pandas_dtype=False):
         whether to infer dtype including pandas extension types.
         If False, scalar belongs to pandas extension types is inferred as
         object
+    downcast : bool, default False
+        If True, downcast float and integer types to the smallest width
+        type that can hold `val`.
+
+    .. versionadded:: 0.20.0
+    allow_uint : bool, default False
+        If True and `downcast` is True, non-negative integers will be
+        downcast to smallest width unsigned integer type that can hold
+        them. Otherwise, signed types are always downcast to signed types
+        and the same for unsigned types.
+
+    .. versionadded:: 0.20.0
     """
 
+    def _downcast_dtype(dtype, val):
+        if 'float' in str(dtype):
+            if ((val > np.finfo(np.float16).min and
+                    val < np.finfo(np.float16).max) or val is np.nan):
+                return np.float16
+            elif (val > np.finfo(np.float32).min and
+                    val < np.finfo(np.float32).max):
+                return np.float32
+            else:
```

Comment:
this would replace your ``_downcast_dtype``; though make it another function in ``pandas.types.cast``

call it ``maybe_downcast_itemsize`` (not to confuse it with the ``maybe_downcast_to_dtype``, which is similar, but doesn't reduce the itemsize)